service.go

```diff
@@ -0,0 +1,541 @@
+		go func(addr ma.Multiaddr) {
+			p, err := addr.ValueForProtocol(ma.P_P2P)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+			pid, err := peer.IDB58Decode(p)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+
+			log.Debugf("pushing record to %s...", p)
+
+			cctx, _ := context.WithTimeout(ctx, reqTimeout)
+			conn, err := s.dial(cctx, pid, grpc.WithInsecure())
+			if err != nil {
+				log.Error(err)
+				return
+			}
+			client := pb.NewThreadsClient(conn)
+			reply, err := client.Push(cctx, req)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+
+			log.Debugf("reply from %s: %t", p, reply.Ok)
+
+			wg.Done()
+		}(addr)
+	}
+
+	// Finally, publish to the thread's topic
+	err = s.publish(id, req)
+	if err != nil {
+		log.Error(err)
+	}
+
+	wg.Wait()
+	return nil
+}
+
+type records struct {
+	sync.RWMutex
+	m map[cid.Cid]thread.Record
+	s []thread.Record
+}
+
+func newRecords() *records {
+	return &records{
+		m: make(map[cid.Cid]thread.Record),
+		s: make([]thread.Record, 0),
+	}
+}
+
+func (r *records) List() []thread.Record {
+	r.RLock()
+	defer r.RUnlock()
+	return r.s
+}
+
+func (r *records) Store(key cid.Cid, value thread.Record) {
+	r.Lock()
+	if _, ok := r.m[key]; ok {
+		return
+	}
+	r.m[key] = value
+	r.s = append(r.s, value)
+	r.Unlock()
+}
+
+// pull records from log addresses.
+func (s *service) pull(ctx context.Context, id thread.ID, lid peer.ID, offset cid.Cid, settings *tserv.PullSettings) ([]thread.Record, error) {
+	lg := s.threads.LogInfo(id, lid)
+	if lg.PubKey == nil {
+		return nil, fmt.Errorf("could not find log")
+	}
+	fk, err := crypto.ParseDecryptionKey(lg.FollowKey)
+	if err != nil {
+		return nil, err
+	}
+
+	req := &pb.PullRequest{
+		Header: &pb.PullRequest_Header{
+			From: &pb.ProtoPeerID{ID: s.threads.host.ID()},
+		},
+		ThreadID: &pb.ProtoThreadID{ID: id},
+		LogID:    &pb.ProtoPeerID{ID: lid},
+		Offset:   &pb.ProtoCid{Cid: offset},
+		Limit:    int32(settings.Limit),
+	}
+
+	// Pull from each address
+	recs := newRecords()
+	wg := sync.WaitGroup{}
+	for _, addr := range s.threads.LogInfo(id, lid).Addrs {
+		wg.Add(1)
+		go func(addr ma.Multiaddr) {
+			p, err := addr.ValueForProtocol(ma.P_P2P)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+			pid, err := peer.IDB58Decode(p)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+
+			log.Debugf("pulling records from %s...", p)
+
+			cctx, _ := context.WithTimeout(ctx, reqTimeout)
+			conn, err := s.dial(cctx, pid, grpc.WithInsecure())
+			if err != nil {
+				log.Error(err)
+				return
+			}
+			client := pb.NewThreadsClient(conn)
+			reply, err := client.Pull(cctx, req)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+
+			for _, r := range reply.Records {
+				rec, err := cbor.RecordFromProto(r, fk)
+				if err != nil {
+					log.Error(err)
+					return
+				}
+				recs.Store(rec.Cid(), rec)
+			}
+
+			log.Debugf("reply from %s: %s", p, reply.String())
+
+			wg.Done()
+		}(addr)
+	}
+
+	wg.Wait()
```

Comment:
As an enhancement, we should probably be emitting these records _live_ via a buffered channel. However, we can revisit when the event store is ready to integrate.